File: Python/Combination Sum.py
#!/usr/bin/env python

import copy

class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def generate(self, candidates, target, seq, cnt, num):
        if target == 0:
            seq.append(copy.copy(cnt))
            return seq
        if num >= len(candidates):
            return seq
        if target >= candidates[num]:
            cnt.append(candidates[num])
            seq = self.generate(candidates, target - candidates[num], seq, cnt, num)
            cnt.pop()
        if num + 1 < len(candidates):
            seq = self.generate(candidates, target, seq, cnt, num+1)
        return seq
    def combinationSum(self, candidates, target):
        seq = []
        cnt = []
        candidates.sort()
        return self.generate(candidates, target, seq, cnt, 0)

File: C++/Combination-Sum.cpp
class Solution {
public:
    void helper(vector<int> &candidates, int target, int index, vector<vector<int> > &res, vector<int> &cur) {
        if (target == 0) {
            res.push_back(cur);
            return;
        }
        for (int i = index; i < candidates.size(); i ++) {
            if (candidates[i] > target) break;
            cur.push_back(candidates[i]);
            helper(candidates, target - candidates[i], i, res, cur);
            cur.pop_back();
        }
    }
    vector<vector<int> > combinationSum(vector<int>& candidates, int target) {
        vector<vector<int> > res;
        sort(candidates.begin(), candidates.end());
        vector<int> cur;
        helper(candidates, target, 0, res, cur);
        return res;
    }
};